sdk/samples/router/general.py

```diff
@@ -0,0 +1,320 @@
+    except TimeoutError:
+        raise Exception(f"Turing API is taking too long for router {my_router.id} to get deployed.")
+
+    # 3. Get the router you just created using the router_id obtained
+    my_router = turing.Router.get(my_router.id)
+    print(f"3. You have retrieved the router with name: {my_router.name}")
+
+    # Access the router config from the returned Router object directly
+    my_router_config = my_router.config
+
+    # Modify something in the router config
+    my_router_config.routes.append(
+        Route(
+            id='fee-fi-fo-fum',
+            endpoint='http://fox-says.fee-fi-fo-fum',
+            timeout='20ms'
+        )
+    )
+
+    # 4. Update the router with the new router config
+    my_router.update(my_router_config)
+    print(f"4. You have just updated your router with a new config.")
+
+    # 5. List all the router config versions of your router
+    my_router_versions = my_router.list_versions()
+    print(f"5. You have just retrieved a list of {len(my_router_versions)} versions for your router:")
+    for ver in my_router_versions:
+        print(ver)
+
+    # Sort the versions returned by version number
+    my_router_versions.sort(key=lambda x: x.version)
+    # Get the version number of the first version returned
+    first_ver_no = my_router_versions[0].version
+    # Get the version number of the latest version returned
+    latest_ver_no = my_router_versions[-1].version
+
+    # Wait for the latest version to get deployed
+    try:
+        my_router.wait_for_version_status(RouterStatus.DEPLOYED, latest_ver_no)
+    except TimeoutError:
+        raise Exception(f"Turing API is taking too long for router {my_router.id} with version {latest_ver_no} to get "
+                        f"deployed.")
+
+    # Wait for the dependencies of the first version to be fully undeployed
+    time.sleep(15)
+
+    # 6. Deploy a specific router config version (the first one we created)
+    response = my_router.deploy_version(first_ver_no)
+    print(f"6. You have deployed version {response['version']} of router {response['router_id']}.")
+
+    # Wait for the first version to get deployed
+    try:
+        my_router.wait_for_version_status(RouterStatus.DEPLOYED, first_ver_no)
+    except TimeoutError:
+        raise Exception(f"Turing API is taking too long for router {my_router.id} with version {first_ver_no} to get "
+                        f"deployed.")
+
+    # 7. Undeploy the current active router configuration
+    response = my_router.undeploy()
+    print(f"7. You have undeployed router {response['router_id']}.")
+
+    # Wait for the router to get undeployed
+    try:
+        my_router.wait_for_status(RouterStatus.UNDEPLOYED)
+    except TimeoutError:
+        raise Exception(f"Turing API is taking too long for router {my_router.id} to get undeployed.")
+
+    # Wait for the dependencies of the router to be fully undeployed
+    time.sleep(10)
```

Comment:
Have you seen some issues if this sleep statement is not added? I think the expectation is that router's status is changed to `UNDEPLOYED` only after all the dependencies are undeployed

Reply:
Yeah I was telling the rest of the team on Thursday that occasionally the deployment of the router/version would take a very long time before failing if it immediately follows an undeployment (without the `sleep` statement). It doesn't occur all time, but somehow spacing out the two actions seems to prevent this from happening. I'll try to see if I can recreate that happening.

Reply:
Ok, as Krithika mentioned, so apparently the deployment failure is caused by the knative podautoscaler encountering some delay in clearing some metrics during the undeployment. 

I suppose in that case it's almost impossible for the client to foresee such a scenario from occurring, so should we just include that sleep statement in the sample or should we build in some safeguards in the methods themselves to prevent these? I was thinking the former would be simpler because it's unlikely most of the users would be switching the same routers on and off endlessly

Reply:
The sleep statements have now been wrapped into the blocking functions in router to hide this away from the clients,